rosetta/transactor/transactor.go

```diff
@@ -0,0 +1,592 @@
+			),
+		}
+	}
+
+	amount := amounts[0]
+	if amount < 0 {
+		amount = -amount
+	}
+
+	intent := Intent{
+		From:     flow.HexToAddress(send.AccountID.Address),
+		To:       flow.HexToAddress(receive.AccountID.Address),
+		Amount:   cadence.UFix64(amount),
+		Payer:    flow.HexToAddress(send.AccountID.Address),
+		Proposer: flow.HexToAddress(send.AccountID.Address),
+		GasLimit: flow.DefaultMaxTransactionGasLimit,
+	}
+
+	return &intent, nil
+}
+
+// CompileTransaction creates a complete Flow transaction from the given intent and metadata.
+func (t *Transactor) CompileTransaction(rosBlockID identifier.Block, intent *Intent, sequence uint64) (string, error) {
+
+	// Generate script for the token transfer.
+	script, err := t.generate.TransferTokens(dps.FlowSymbol)
+	if err != nil {
+		return "", fmt.Errorf("could not generate transfer script: %w", err)
+	}
+
+	// TODO: Allow arbitrary proposal key index
+	// => https://github.com/optakt/flow-dps/issues/369
+
+	// Create the transaction.
+	unsignedTx := sdk.NewTransaction().
+		SetScript(script).
+		SetReferenceBlockID(sdk.HexToID(rosBlockID.Hash)).
+		SetPayer(sdk.Address(intent.Payer)).
+		SetProposalKey(sdk.Address(intent.Proposer), 0, sequence).
+		AddAuthorizer(sdk.Address(intent.From)).
+		SetGasLimit(intent.GasLimit)
+
+	receiver := cadence.NewAddress(flow.BytesToAddress(intent.To.Bytes()))
+
+	// Add the script arguments - the amount and the receiver.
+	_ = unsignedTx.AddArgument(intent.Amount)
+	_ = unsignedTx.AddArgument(receiver)
+
+	payload, err := t.encodeTransaction(unsignedTx)
+	if err != nil {
+		return "", fmt.Errorf("could not encode transaction: %w", err)
+	}
+
+	return payload, nil
+}
+
+func (t *Transactor) HashPayload(rosBlockID identifier.Block, unsigned string, signer identifier.Account) (string, error) {
+
+	unsignedTx, err := t.decodeTransaction(unsigned)
+	if err != nil {
+		return "", fmt.Errorf("could not decode transaction: %w", err)
+	}
+
+	// Validate block.
+	height, _, err := t.validate.Block(rosBlockID)
+	if err != nil {
+		return "", fmt.Errorf("could not validate block: %w", err)
+	}
+
+	// Validate address.
+	address, err := t.validate.Account(signer)
+	if err != nil {
+		return "", fmt.Errorf("could not validate account: %w", err)
+	}
+
+	key, err := t.invoke.Key(height, address, 0)
+	if err != nil {
+		return "", fmt.Errorf("could not retrieve key: %w", err)
+	}
+
+	message := unsignedTx.EnvelopeMessage()
+	message = append(flow.TransactionDomainTag[:], message...)
+
+	hasher, err := crypto.NewHasher(key.HashAlgo)
+	if err != nil {
+		return "", fmt.Errorf("could not create hasher: %w", err)
+	}
+
+	hash := hex.EncodeToString(hasher.ComputeHash(message))
+
+	return hash, nil
+}
+
+// ParseTransactions processes the flow transaction, validates its correctness and translates it
+// to a list of operations and a list of signers.
+func (t *Transactor) ParseTransaction(payload string) (identifier.Block, uint64, []object.Operation, []identifier.Account, error) {
```

Comment:
That's a lot of return values!!

<img src="https://i.giphy.com/media/qWnxA0E2LmcMM/giphy.gif?cid=ecf05e47do21df43en6d0gitn0of1mljcpuk23v8isce5nci&rid=giphy.gif&ct=g" />

I'd suggest returning `(*transactor.Transaction, error)` which contains all those things. 🙏  🙏  🙏  🙏

Reply:
I don't currently like this function, but it's the correct approach if we respect the separation of responsibilities between the Rosetta API package and the Rosetta business logic. The "native" Rosetta types should be used exclusively, which are the `identifier`, `object` and `meta` structs, `int` for key index, `uint64` for sequence number and `string` for base64-encoded JSON-marshalled transaction payloads.

Reply:
The refactoring solution works for me. Will we do it before merging this PR?

Reply:
No, due to time pressure.